### PR TITLE
Refactor application deploy

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Refactor application deploy ahead of charm hub changes.  These changes also allow for more detailed mock tests of each section.  Not all tests are mocked yet.

Most of deploy can be separated out into a deployer package.  It has a factory to return the type of deploy being done:
- local charm
- charm store charm
- local bundle
- charm store bundle 

BuildModelRepresentation is moved to a new bundle package.  

The store package handles charm and bundle URL resolution.

## QA steps

Any and every deploy and bundle you can think of, including local/charmstore, bundles, overlays, existing machines, resources.

### CODE REVIEW

**Comment by hmlanigan** [comment]
force push to rebase

**Comment by hmlanigan** [comment]
!!build!!

**Comment by hmlanigan** [comment]
failed on intermittent: localTests.TestPortsToRuleInfo

Already fixed in 2.8 branch

File: cmd/juju/application/deployer/charm.go [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+type predeployedLocalCharm struct {
+	deployCharm
+	userCharmURL *charm.URL
+}
+
+// PrepareAndDeploy finishes preparing to deploy a predeployed local charm,
+// then deploys it.
+func (d *predeployedLocalCharm) PrepareAndDeploy(ctx *cmd.Context, deployAPI DeployerAPI, _ *store.CharmStoreAdaptor) error {
+	userCharmURL := d.userCharmURL
+	ctx.Verbosef("Preparing to deploy local charm %q again", userCharmURL.Name)
```
Comment (SimonRichardson) [inline cmd/juju/application/deployer/charm.go:262]: `again`?

File: cmd/juju/application/deployer/deployer.go [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+type Deployer interface {
+	// PrepareAndDeploy finishes preparing to deploy a charm or bundle,
+	// then deploys it.  This is done as one step to accomidate the
+	// call being wrapped by block.ProcessBlockedError.
+	PrepareAndDeploy(*cmd.Context, DeployerAPI, *store.CharmStoreAdaptor) error
+}
+
+// GetDeployer returns the correct deployer to use based on the cfg provided.
+// A ModelConfigGetter and CharmStoreAdaptor needed to find the deployer.
+func (d *factory) GetDeployer(cfg DeployerConfig, getter ModelConfigGetter, cstoreAPI *store.CharmStoreAdaptor) (Deployer, error) {
```
Comment (SimonRichardson) [inline cmd/juju/application/deployer/deployer.go:68]: Lovely.

File: cmd/juju/application/deployer/deployer.go [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+type DeployerDependencies struct {
+	DeployResources      resourceadapters.DeployResourcesFunc
+	Model                ModelCommand
+	NewConsumeDetailsAPI func(url *charm.OfferURL) (ConsumeDetails, error)
+	Steps                []DeployStep
+}
+
+// DeployerConfig is the data required to choose a deployer and then run
+// PrepareAndDeploy.
+type DeployerConfig struct {
```
Comment (SimonRichardson) [inline cmd/juju/application/deployer/deployer.go:123]: This is terrible, but I don't see how to fix this in the short term.

Can we add a TODO to refactor this config into a better structure?

File: cmd/juju/application/deployer/interface.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+)
+
+// DeployStepAPI represents a API required for deploying using the step
+// deployment code.
+type DeployStepAPI interface {
+	MeteredDeployAPI
+}
+
+// DeployStep is an action that needs to be taken during charm deployment.
+type DeployStep interface {
```
Comment (SimonRichardson) [inline cmd/juju/application/deployer/interface.go:32]: Shame this abstraction isn't used more extensively.

File: cmd/juju/application/store/interface.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+type CharmAdder interface {
+	AddLocalCharm(*charm.URL, charm.Charm, bool) (*charm.URL, error) // not used in utils
+	AddCharm(*charm.URL, csparams.Channel, bool) error
+	AddCharmWithAuthorization(*charm.URL, csparams.Channel, *macaroon.Macaroon, bool) error
+}
+
+// charmrepoForDeploy is a stripped-down version of the
+// gopkg.in/juju/charmrepo.v4 Interface interface. It is
+// used by tests that embed a DeploySuiteBase.
+type CharmrepoForDeploy interface {
```
Comment (SimonRichardson) [inline cmd/juju/application/store/interface.go:23]: I think this is superfluous for charmhub, but need to delve deeper.

File: cmd/juju/application/store/store.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2012, 2013 Canonical Ltd.
```
Comment (SimonRichardson) [inline cmd/juju/application/store/store.go:1]: Love this in in a store package, we should be able to do something for charmhub, so this is perfect naming.

File: cmd/juju/application/utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+// ParsePlacement validates provided placement of a unit and
+// returns instance.Placement.
+func ParsePlacement(spec string) (*instance.Placement, error) {
+	if spec == "" {
+		return nil, nil
+	}
+	placement, err := instance.ParsePlacement(spec)
+	if err == instance.ErrPlacementScopeMissing {
+		spec = "model-uuid" + ":" + spec
```
Comment (SimonRichardson) [inline cmd/juju/application/utils/utils.go:37]: I know you didn't write this, but this makes my eyes bleed:


```suggestion
		spec = fmt.Sprintf("model-uuid:%s", spec)
```

File: cmd/juju/application/bundle/bundle_test.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+	bundleDataSource *mocks.MockBundleDataSource
+}
+
+var _ = gc.Suite(&composeAndVerifyRepSuite{})
+
+func (s *composeAndVerifyRepSuite) TestComposeAndVerifyBundleNoOverlay(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+	_, err := ComposeAndVerifyBundle(s.bundleDataSource, nil)
+	c.Assert(err, jc.ErrorIsNil)
+	// TODO Verify result
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply cmd/juju/application/bundle/bundle_test.go]: This was fixed in commit 083eef3, a few more tests added too.

File: cmd/juju/application/deployer/deployer.go (hunk 3 above)
Comment (hmlanigan) [reply cmd/juju/application/deployer/deployer.go:123]: adding

File: cmd/juju/application/deployer/deployer_test.go [hunk 9]
```diff
@@ -0,0 +1,4 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package deployer
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply cmd/juju/application/deployer/deployer_test.go:4]: added.

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**